File: ros1_to_ros2_migration_tutorial/ros2_app_cpp/src/number_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/int64.hpp"

class NumberPublisherNode : public rclcpp::Node
{
    public:
        NumberPublisherNode(): Node("number_publisher")
        {
            this->declare_parameter("number_to_publish", 2);
            this->declare_parameter("number_publish_frequency", 1.0);

            number_ = this->get_parameter("number_to_publish").as_int();
            double publish_frequency = this->get_parameter("number_publish_frequency").as_double();

            number_publisher_ = this->create_publisher<std_msgs::msg::Int64>("number", 10);

            number_timer_ = this->create_wall_timer(std::chrono::milliseconds((int)(1000.0 / publish_frequency)),
                                                    std::bind(&NumberPublisherNode::publishNumber, this));
            RCLCPP_INFO(this->get_logger(), "Number publisher has been started");
        }

    private:
        void printNumber()
        {
            RCLCPP_INFO(this->get_logger(), "%d", number_);
        }

        void publishNumber()
        {
            // auto msg = std_msgs::msg::Int64();
            std_msgs::msg::Int64 msg;
            msg.data = number_;
            number_publisher_->publish(msg);
        }

        int number_;
        rclcpp::Publisher<std_msgs::msg::Int64>::SharedPtr number_publisher_;
        rclcpp::TimerBase::SharedPtr number_timer_;
};

int main(int argc, char ** argv)
{
    rclcpp::init(argc, argv);  // init ROS 2 communication (no node created here)

    // create a node object, stored in a shared pointer
    // auto node = std::make_shared<rclcpp::Node>("number_publisher");  // basic way
    auto node = std::make_shared<NumberPublisherNode>();  // OOP way

    rclcpp::spin(node);

    rclcpp::shutdown();  // shutdown ROS 2 communication

    return 0;
}

File: ros1_to_ros2_migration_tutorial/ros2_app_py/ros2_app_py/number_publisher.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from std_msgs.msg import Int64

class NumberPublisherNode(Node):
    def __init__(self):
        super().__init__("number_publisher")
        self.number_ = 4

        self.pub_ = self.create_publisher(Int64, "number", 10)

        self.number_timer_ = self.create_timer(0.5, self.publish_number)
        self.get_logger().info("Number publisher has been started!!!")

    def print_number(self):
        self.get_logger().info(str(self.number_))

    def publish_number(self):
        msg = Int64()
        msg.data = self.number_
        self.pub_.publish(msg)

def main(args=None):
    rclpy.init(args=args)  # init ROS2 communication (no node created here)

    ## node = Node('number_publisher')  # ROS2 node creation - basic way

    node = NumberPublisherNode()  # ROS2 node creation - OOP way

    rclpy.spin(node)  # note that we pass to spin() the name of the node we want to spin --> in ROS2 a script can contain multiple nodes

    rclpy.shutdown()  # shutdown ROS2 communication

if __name__ == '__main__':
    main()